### DISCUSSION TITLE
Make the `HttpContextContract` a generic interface

### DISCUSSION THREAD

**Discussion by ndianabasi** [adonisjs/core#4210] [Ideas]
I've felt handicapped by the inability to type the `params` context property all through my journey with `V5`. It would wonderful if `V6` makes the `HttpContextContract` interface a generic one so that the `params` property by the typed by the developer.

Example:

```ts
type Params = {
  user_id?: string
  organisation_id?: string
}

export default class ExampleController {
  public async store({ request, bouncer, params }: HttpContextContract<Params>) {
    // Strongly-typed `organisation_id`
    if (params.organisation_id) {
    }

    // Strongly-typed `user_id`
    if (params.user_id) {
    }
  }
}
```

**Comment by ndianabasi** [comment]
I just tried this and it worked:

```ts
type Params = {
  user_id?: string
  organisation_id?: string
}

export default class ExampleController {
                                           // Using type intersection 👇
  public async store({ request, bouncer, params }: HttpContextContract & { params: Params }) {
    // Strongly-typed `organisation_id`
    if (params.organisation_id) {
    }

    // Strongly-typed `user_id`
    if (params.user_id) {
    }
  }
}
```

However, using a generic interface is neater.

**Comment by RomainLanz** [comment]
In my opinion, it would be better to use assertion to ensure your params are defined.

Using generic here does not bring any validation or security.

**Reply by samueldrai** [reply]
Using Vine or Zod, it would allow for efficient request body/qs/params validation I think. 
@RomainLanz is there a way to hack around it? :) 

You could pass a schema to your route using a middleware and the `request.body()` function would return an object whose type would be inferred from the schema.
